### ISSUE TITLE
Edits are broken for serialized session layers or SdfLayers that cannot handle edits.

### ISSUE THREAD

**Issue by marsupial** [Autodesk/maya-usd#898]
**Describe the bug**
Maya is now setting the stage's root layer to the edit target by default.
This doesn't work for layer's that are not editable.
Additionally, during `MayaUsdProxyShapeBase::computeInStageDataCached` a session layer can be delivered via `computeSessionLayer` and the layer opened, but then Maya will immediately move edits into the stage's root-layer.

**Comment by marsupial** [comment]
https://github.com/Autodesk/maya-usd/pull/899

**Comment by wtelford** [comment]
Hey,
We would love to understand the workflows that you are leveraging here. The changes made internally were to

**Comment by wtelford** [comment]
We would love to understand the workflows that you are trying to leverage here. 
The changes made internally were to:
1. Change the default edit target to the root layer for USD files that are opened.
2. Hide the session layer in the layer editor if the session layer is empty or not targeted (with the option to disable this behavior).
The session layer is always available for targeting in either case.
I am also not currently aware of anything in MayaUSD setting layers as not editable yet. I'm assuming you are referring to setting "PermissionToEdit" on a layer. When we do leverage that feature we will take care to make sure we also change our target when we do, as that will be a problem regardless of whether we are talking about the root layer or any other layer in the stack.

**Comment by marsupial** [comment]
So the most important use case right now is that we have FileFormat plug-ins which are "read-only", so when you're setting the EditTarget to those layers, it will constantly error out.  That's the point of checking after load, we want to make sure edit's are allowed, but we definitely cannot target the input format's SdfLayer.

At a minimum we'd like to be able to move to a session layer 'automatically' for formats that one cannot author to, so that they are not locked out of changes.  I don't think the PR proposed should interfere with whatever you'd want to do in the future with the SdfLayer permission API (i.e. user controllable locking etc) as that field is never serialized, so you'll essentially have to author some metadata that Maya will need to read after loading to know if it should "Lock the Layer" after load, AFAIK the only way `PermissionToEdit` will return false immediately after SdfLayer::Load, is that the underlying format has explicitly set it for the SdfLayer.

Make sense ?

**Comment by kxl-adsk** [comment]
Thank you @marsupial. I have a question about the second part of your change, i.e. 
> Additionally, during MayaUsdProxyShapeBase::computeInStageDataCached a session layer can be delivered via computeSessionLayer and the layer opened, but then Maya will immediately move edits into the stage's root-layer.

Based on the title of this issue, I'm assuming you are using this for the session layer serialization. I other words, you are proposing that we should target the session layer if it was retrieved for a Maya file during the file open - did I got it right?

And just to avoid future code conflicts - Is your session layer serialization code leveraging #871 or you wrote your own?

**Comment by marsupial** [comment]
We have nothing custom in this arena (outside of this PR), and would like it to _work out of the box_.

Generally in the example above where we require a session layer to do edits: re-targeting to that session-layer on re-open (or import) is important.

**Comment by kxl-adsk** [comment]
With pull-request #661, we made a deliberate change to target the stage root layer for every proxy shape. You brought up an excellent case where the generated layer by (dynamic) file format plugin won't have permission to edit. We haven't considered this before and agree with the proposal to fallback to the session layer. 

Separately, you are proposing an additional change (based on the pull-request you opened, i.e. #899). Every proxy shape that implements the computeSessionLayer method will target the returned session layer instead of the root layer (https://github.com/marsupial/maya-usd/blob/PR/898-restore-editability/lib/mayaUsd/nodes/proxyShapeBase.cpp#L553). In other words, your proposal effectively reverts our change for:
- Pixar proxy shape (it implements computeSessionLayer) 
- For Autodesk proxy shape saved in Maya's file (behavior after PR #871)

We are not keen on creating such differences between proxy shapes as it could result in problems for companies to migrate from Pixar proxy shape to Autodesk proxy shape. Instead, would it solve your problems if we have a fall back to the legacy behavior where we target the session layer? (this could be an env variable or option var)

So to summarize, we propose to target the session layer if
- root layer is not permitting edits, or
- a fall back behavior is enabled

**Comment by marsupial** [comment]
I also think consistency is prob paramount for switching, so as long as a "un-editable root layer" gets an editable default and enabling fallback behavior is a real user-options (not some ENV_VAR), this works for me!

**Comment by scottrenaud** [comment]
Hey @marsupial,

We posted [Feedback needed on new UI - Layer Locking #3486](https://github.com/Autodesk/maya-usd/discussions/3486) and would love your feedback on the designs if you have a chance!  It relates to what you discussed in this thread (a while ago) and would be great to validate that we're on the right track.

Thanks!
Scott
